aws_lightsail_sd

n888 opened the pull request:
Hello @roidelapluie & Team,

Some notes:

- discovery/ec2 has been moved to discovery/amazon and merged with lightsail code (wondering if the naming/capitalizations are fit)
- I did not see any pagination "Pages" API related function for `GetInstancesWithContext`, also `GetInstancesInput` does not support a Filter, so I have not included any logic there. I assume these are missing because the lightsail instance limits are low (20/region).

Thank you in advance for your time and reviewing.

roidelapluie commented:
Thank you! I will review the PR soon.

roidelapluie reviewed (commented):
I have taken a first look. What do you think about moving lightsail in its own go file? Now this is interleaving ec2 and lightsail which is not really nice. e.g. `aws/aws.go aws/ec2.go aws/lightsail.go `?


You have forgotten the documentation.

roidelapluie commented:
I have renamed dockerswarm to moby so you mught want to rebase this PR. What do you think about `aws` as package name BTW?

n888 commented:
Docs added, also added missing RoleARN code and split out ec2.go/lightsail.go.
I think `aws` is the better package name, as it is shorter, and seems to be more widely referenced.

n888 commented on discovery/amazon/amazon.go:
```diff
@@ -298,3 +420,70 @@ func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+
+	output, err := lightsailClient.GetInstancesWithContext(ctx, input)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && (awsErr.Code() == "AuthFailure" || awsErr.Code() == "UnauthorizedOperation") {
```
Please excuse my (beginner) dev skills, can you help me understand what consts I should be looking for?

roidelapluie commented on discovery/amazon/ec2.go:
```diff
@@ -61,24 +61,26 @@ const (
-}
+var (
+	// EC2DefaultSDConfig is the default EC2 SD configuration.
+	EC2DefaultSDConfig = EC2SDConfig{
```
```suggestion
	DefaultEC2SDConfig = EC2SDConfig{
```

n888 replied:
Updated to `DefaultEC2SDConfig` & `DefaultLightsailSDConfig`

roidelapluie commented on discovery/amazon/lightsail.go:
```diff
@@ -0,0 +1,235 @@
+// Copyright 2015 The Prometheus Authors
```
```suggestion
// Copyright 20121The Prometheus Authors
```

n888 replied:
Updated copyright in `ec2.go` & `lightsail.go`

roidelapluie commented on discovery/amazon/lightsail.go:
```diff
@@ -0,0 +1,235 @@
+
+var (
+	// LightsailDefaultSDConfig is the default Lightsail SD configuration.
+	LightsailDefaultSDConfig = LightsailSDConfig{
```
```suggestion
	DefaultLightsailSDConfig = LightsailSDConfig{
```

n888 replied:
Updated

roidelapluie commented on discovery/amazon/lightsail.go:
```diff
@@ -0,0 +1,235 @@
+
+	output, err := lightsailClient.GetInstancesWithContext(ctx, input)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && (awsErr.Code() == "AuthFailure" || awsErr.Code() == "UnauthorizedOperation") {
```
Are there consts for this?

roidelapluie replied:
I have checked and could not find that so that is OK

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -238,6 +238,10 @@ hetzner_sd_configs:
   [ - <kubernetes_sd_config> ... ]
 
+# List of Lightsail service discovery configurations.
+lightsail_sd_configs:
```
This also needs to go to the alerting SD section

n888 replied:
https://github.com/prometheus/prometheus/pull/8693/commits/ff78f97d5139e5c9f4b171c7536c9e3ec11d840c

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1341,6 +1345,53 @@ for a detailed example of configuring Prometheus for Kubernetes.
 
+### `<lightsail_sd_config>`
+
+Lightsail SD configurations allow retrieving scrape targets from AWS Lightsail
```
Can you add a link to lightsail?

n888 replied:
https://github.com/prometheus/prometheus/pull/8693/commits/3db49731a807b670e91fc84ce6f402f19cb8a0df

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1341,6 +1345,53 @@ for a detailed example of configuring Prometheus for Kubernetes.
+* `__meta_lightsail_instance_state`: the state of the Lightsail instance
+* `__meta_lightsail_instance_support_code`: the support code of the Lightsail instance
+* `__meta_lightsail_ipv6_addresses`: comma separated list of IPv6 addresses assigned to the instance's network interfaces, if present
+* `__meta_lightsail_private_ip`: the private IP address of the instance, if present
```
```suggestion
* `__meta_lightsail_private_ip`: the private IP address of the instance
```

We mandate that

n888 replied:
https://github.com/prometheus/prometheus/pull/8693/commits/ffdb9a687b15ea33962887a9110e610986c8bd99

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1341,6 +1345,53 @@ for a detailed example of configuring Prometheus for Kubernetes.
+
+# The port to scrape metrics from. If using the public IP address, this must
+# instead be specified in the relabeling rule.
+[ port: <int> | default = 80 ]
```
You have forgotten to close the code block (with <code>```</code>)

n888 replied:
https://github.com/prometheus/prometheus/pull/8693/commits/1981dc9c84393bb09e1ecffaf43b2dc10d4a817e

roidelapluie commented on discovery/amazon/lightsail.go:
```diff
@@ -0,0 +1,235 @@
+		}
+
+		labels := model.LabelSet{
+			lightsailLabelInstanceSupportCode: model.LabelValue(*inst.SupportCode),
```
Can we set the other here? like IP,  blueprint etc..

n888 replied:
https://github.com/prometheus/prometheus/pull/8693/commits/45b8e9f303661f48eda638d9ad9fee65ee6aec95
Do we want any more labels inside here?

roidelapluie commented:
Thank you. You have used amzon as name, can you switch to AWS? You will also need to fix the DCO and run go mod tidy for CI.

n888 commented:
> Thank you. You have used amzon as name, can you switch to AWS? You will also need to fix the DCO and run go mod tidy for CI.

Much appreciated,

 - aws rename -> https://github.com/prometheus/prometheus/pull/8693/commits/73fe6450d2a2193fd95792e561f0ebc698f5ca2b

roidelapluie reviewed (approved):
Thank you, this is ready to be merged once you fix the tests (run go mod tidy I think)

n888 commented:
Thanks, so I got the go mod tidy sorted by upgrading to go 1.16.3. Any idea what's going on with `ci/circleci: test_windows` ?

roidelapluie commented:
Thank you!!!!!
